erpnext/accounts/doctype/payment_manager/payment_manager.py

```diff
@@ -0,0 +1,429 @@
+
+			query = """select posting_date, name, {0}, outstanding_amount
+							from `tab{1}` where ROUND(outstanding_amount)={2} and posting_date < '{3}'
+							""".format(party_type, inv_type, round(abs(entry.amount)), entry.transaction_date)
+			invoices = frappe.db.sql(query, as_dict = True)
+			if(len(invoices) > 0):
+				entry.party = invoices[0].get(party_type)
+
+	def populate_matching_vouchers(self):
+		for entry in self.new_transaction_items:
+			if (not entry.party or entry.reference_name): continue
+			print("Finding matching voucher for {0}".format(entry.description))
+			amount = abs(entry.amount)
+			invoices = []
+			vouchers = get_matching_journal_entries(self.from_date, self.to_date, entry.party, self.bank_account, amount)
+			if len(vouchers) == 0: continue
+			for voucher in vouchers:
+				added = next((entry.invoice for entry in self.payment_invoice_items if entry.invoice == voucher.voucher_no), None)
+				if (added):
+					print("Found voucher {0}".format(added))
+					continue
+				print("Adding voucher {0} {1} {2}".format(voucher.voucher_no, voucher.posting_date, voucher.debit))
+				ent = self.append('payment_invoice_items', {})
+				ent.invoice_date = voucher.posting_date
+				ent.invoice_type = "Journal Entry"
+				ent.invoice = voucher.voucher_no
+				ent.payment_description = entry.description
+				ent.allocated_amount = max(voucher.debit, voucher.credit)
+
+				invoices += [ent.invoice_type + "|" + ent.invoice]
+				entry.reference_type = "Journal Entry"
+				entry.mode_of_payment = "Wire Transfer"
+				entry.reference_name = ent.invoice
+				#entry.account = entry.party
+				entry.invoices = ",".join(invoices)
+				break
+
+
+	def create_payment_entries(self):
+		for payment_entry in self.new_transaction_items:
+			if (not payment_entry.party): continue
+			if (payment_entry.reference_name): continue
+			print("Creating payment entry for {0}".format(payment_entry.description))
+			if (payment_entry.party_type == "Account"):
+				payment = self.create_journal_entry(payment_entry)
+				invoices = [payment.doctype + "|" + payment.name]
+				payment_entry.invoices = ",".join(invoices)
+			else:
+				payment = self.create_payment_entry(payment_entry)
+				invoices = [entry.reference_doctype + "|" + entry.reference_name for entry in payment.references if entry is not None]
+				payment_entry.invoices = ",".join(invoices)
+				payment_entry.mode_of_payment = payment.mode_of_payment
+				payment_entry.account = self.receivable_account if payment_entry.party_type == "Customer" else self.payable_account
+			payment_entry.reference_name = payment.name
+			payment_entry.reference_type = payment.doctype
+		frappe.msgprint(_("Successfully created payment entries"))
+
+	def create_payment_entry(self, pe):
+		payment = frappe.new_doc("Payment Entry")
+		payment.posting_date = pe.transaction_date
+		payment.payment_type = "Receive" if pe.party_type == "Customer" else "Pay"
+		payment.mode_of_payment = "Wire Transfer"
```

Comment:
hard code?

Reply:
there are many, which in particular? 
I chose 'Wire Transfer' as all the entries are direct bank transfers.  I intended to make it configurable, but not sure managed to get that. 
Remaining ones like "Customer", "Pay" and "Receive" are standards, so not sure whether there is a better way of doing it.